File: tools/find_missing.py
#!/usr/bin/env python3
"""Find all missing translations."""
import os
import re

base_path = os.path.join(os.path.dirname(__file__), '..', 'app', 'src', 'main', 'res')
en_file = os.path.join(base_path, 'values', 'strings.xml')

with open(en_file, 'r', encoding='utf-8') as f:
    en_content = f.read()

# Get all English string names and values
en_strings = {}
for match in re.finditer(r'<string name="([^"]+)"[^>]*>([^<]+)</string>', en_content):
    en_strings[match.group(1)] = match.group(2)

output_lines = []
output_lines.append(f'Total English strings: {len(en_strings)}')
output_lines.append('')

# Check each language
lang_dirs = [d for d in os.listdir(base_path) if d.startswith('values-') and os.path.isdir(os.path.join(base_path, d))]
lang_dirs.sort()

all_missing = {}

for lang_dir in lang_dirs:
    lang_file = os.path.join(base_path, lang_dir, 'strings.xml')
    if os.path.exists(lang_file):
        with open(lang_file, 'r', encoding='utf-8') as f:
            lang_content = f.read()

        # Get translated strings (where value differs from English)
        lang_strings = {}
        for match in re.finditer(r'<string name="([^"]+)"[^>]*>([^<]+)</string>', lang_content):
            lang_strings[match.group(1)] = match.group(2)

        # Count translated (different from English)
        translated = 0
        untranslated = []
        for name, en_val in en_strings.items():
            if name in lang_strings:
                if lang_strings[name] != en_val:
                    translated += 1
                else:
                    untranslated.append(name)
            else:
                untranslated.append(name)

        missing = len(en_strings) - translated
        pct = (translated / len(en_strings)) * 100 if len(en_strings) > 0 else 0

        all_missing[lang_dir] = untranslated

        status = 'OK' if pct >= 50 else 'NEED'
        bar = '#' * int(pct / 5) + '-' * (20 - int(pct / 5))
        output_lines.append(f'{status} {lang_dir:12} [{bar}] {pct:5.1f}% ({translated}/{len(en_strings)}) - {missing} need work')

output_lines.append('')

# Show sample missing strings for worst languages
worst = sorted(all_missing.items(), key=lambda x: len(x[1]), reverse=True)[:5]
output_lines.append("Sample missing strings from languages needing most work:")
for lang, missing in worst:
    output_lines.append(f"\n{lang}: {len(missing)} untranslated")
    for s in missing[:10]:
        output_lines.append(f"  - {s}")

# Write to file
output_file = os.path.join(os.path.dirname(__file__), 'trans_status.txt')
with open(output_file, 'w', encoding='utf-8') as f:
    f.write('\n'.join(output_lines))

print("Output written to trans_status.txt")

File: tools/find_missing.js
const fs = require('fs');
const path = require('path');

const resDir = path.join(__dirname, '..', 'app', 'src', 'main', 'res');
const outFile = path.join(__dirname, 'missing_keys.txt');

function extractKeys(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    const keys = [];
    const regex = /name="([^"]+)"/g;
    let match;
    while ((match = regex.exec(content)) !== null) {
        keys.push(match[1]);
    }
    return keys;
}

function extractStringEntries(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    const entries = {};
    // Match <string name="key">value</string> (including multiline)
    const regex = /<string\s+name="([^"]+)"[^>]*>([\s\S]*?)<\/string>/g;
    let match;
    while ((match = regex.exec(content)) !== null) {
        entries[match[1]] = match[2];
    }
    return entries;
}

const baseFile = path.join(resDir, 'values', 'strings.xml');
const baseKeys = extractKeys(baseFile);
const baseEntries = extractStringEntries(baseFile);

let output = `Base English has ${baseKeys.length} entries\n\n`;

const langs = ['tr', 'es', 'de', 'fr', 'pt'];
const allMissing = {};

for (const lang of langs) {
    const langFile = path.join(resDir, `values-${lang}`, 'strings.xml');
    const langKeys = new Set(extractKeys(langFile));

    const missing = baseKeys.filter(k => !langKeys.has(k));
    allMissing[lang] = missing;

    output += `=== ${lang} === ${langKeys.size}/${baseKeys.length} (missing: ${missing.length})\n`;
    for (const k of missing) {
        const engVal = baseEntries[k] || '(non-string entry)';
        output += `  ${k} = ${engVal}\n`;
    }
    output += '\n';
}

fs.writeFileSync(outFile, output, 'utf8');
console.log('Done. Output written to ' + outFile);

